Source/WebKit/Platform/cocoa/CocoaHelpers.mm

```diff
@@ -186,6 +188,92 @@
     });
 }
 
+// MARK: NSDictionary helper methods.
+
+NSDictionary *dictionaryWithLowercaseKeys(NSDictionary *dictionary)
+{
+    if (!dictionary.count)
+        return @{ };
+
+    NSMutableDictionary *newDictionary = [NSMutableDictionary dictionaryWithCapacity:dictionary.count];
+    for (NSString *key in dictionary.allKeys) {
+        if (![key isKindOfClass:NSString.class]) {
+            ASSERT_NOT_REACHED();
+            continue;
+        }
+
+        newDictionary[key.lowercaseString] = dictionary[key];
+    }
+
+    return [newDictionary copy];
+}
+
+NSDictionary *mergeDictionaries(NSDictionary *dictionaryA, NSDictionary *dictionaryB)
+{
+    if (!dictionaryB.count)
+        return dictionaryA;
+
+    if (!dictionaryA.count)
+        return dictionaryB;
+
+    NSMutableDictionary *mergedDictionary = [dictionaryA mutableCopy];
+
+    for (id key in dictionaryB.allKeys) {
+        if (!dictionaryA[key])
+            mergedDictionary[key] = dictionaryB[key];
+    }
+
+    return mergedDictionary;
+}
+
+NSDictionary *mergeDictionariesAndSetValues(NSDictionary *dictionaryA, NSDictionary *dictionaryB)
+{
+    if (!dictionaryB.count)
+        return dictionaryA;
+
+    if (!dictionaryA.count)
+        return dictionaryB;
+
+    NSMutableDictionary *newDictionary = [dictionaryA mutableCopy];
+
+    for (id key in dictionaryB.allKeys)
+        newDictionary[key] = dictionaryB[key];
+
+    return [newDictionary copy];
+}
+
+// MARK: NSLocale helper methods.
+
+NSString *localeStringInWebExtensionFormat(NSLocale *locale)
+{
+    if (!locale.languageCode)
+        return @"";
+
+    if (locale.countryCode.length)
+        return [NSString stringWithFormat:@"%@-%@", locale.languageCode, locale.countryCode];
+    return locale.languageCode;
+}
+
+// MARK: NSError helper methods
+
+NSString *privacyPreservingDescription(NSError *error)
+{
+    NSString *privacyPreservingDescription = objectForKey<NSString>(error.userInfo, WebkitPrivacyPreservingDescription);
+    if (!privacyPreservingDescription) {
+        NSString *domain = error.domain;
+        if (domain.length) {
+            id (^valueProvider)(NSError *err, NSString *userInfoKey) = [NSError userInfoValueProviderForDomain:domain];
+            if (valueProvider)
+                privacyPreservingDescription = valueProvider(error, WebkitPrivacyPreservingDescription);
+        }
+    }
+
+    if (privacyPreservingDescription)
+        return [NSString stringWithFormat:@"Error Domain=%@ Code=%ld \"%@\"", error.domain, (long)error.code, privacyPreservingDescription];
+
+    return [NSError errorWithDomain:error.domain ?: @"" code:error.code userInfo:nil].description;
+}
+
```

Comment:
I learned something new recently. NSErrors will be redacted automatically by `os_log` when used in a plain `%@`  or `%{private}@` placeholder by calling the `redactedDescription` method on the `NSError` object (and many other objects). We don't need to do anything like this anymore!

Reply:
I also don'y see you using this anywhere, so I would just remove it for now. I think logging the error as `%{private}@` is fine for now.

Reply:
This is being used on line 82 in `_WKWebExtensionLocalization.mm`